### ISSUE TITLE
OpenCv 3.4.1 for Android: Bad JavaCamera2View performance 

### ISSUE THREAD

**Issue by ansanders** [opencv/opencv#11261]
#I recently upgraded to 3.4.1 and build it to support new `JavaCamera2View`. 
On my Nexus 7 (2013) I resolved a FPS decrease over time bug with `JavaCameraView`. 
`JavaCamera2View` now gives slow, but stable ~10 FPS . Not acceptable but maybe improvable. 
Performance on my Galaxy S8 differs much. ~30 FPS with `JavaCameraView`, without any processing. 
But only **2** FPS! with `JavaCamera2View`, which is not usable for any needs. 
In both test cases, `JavaCamera2View` has a very improved behaviour in resolution settings, scaling / zooming behaviour and image quality (removed some blurry and bad quality effects) and maybe that is some kind of the problem. 
As far as I can say the resolution, quality and autofocus behaviour is so good, that it drops the FPS to a minimum.  But that is just my "newbie" opinion. 
Buffer managment could also be a problem, but I can't imagine why the performance is better on an older Device than S8 (Nexus 7). 

really need to solve this issue because I'd love to use the new `JavaCamera2View` since the quality of camera is much better

ty in advance

##### System information (version)

- OpenCV => 3.4.1 with contrib
- Operating System / Platform => Ubuntu 16.04
- Compiler Settings => build_sdk.py with ndk-config.16 - downgraded android SDK 

#### Edit

the devices get really really hot after a few minutes of running a simple Java2CameraView

**Comment by ansanders** [comment]
@floe let me know if u need more details about my setup or something to solve this nasty little bug

**Comment by ansanders** [comment]
linked to [https://github.com/opencv/opencv/issues/11118](url) of **tipsmiller ** " replacing the Imgproc.COLOR_YUV2RGBA_NV21 value with Imgproc.COLOR_YUV420sp2BGR" worked for me too with the color behaviour bug

**Comment by floe** [comment]
@ansanders thanks for the thorough analysis. I think the root cause is that `calcPreviewSize` is still basically the same code as for the original Camera1 API, however, that was much more limited in what preview sizes it supported (at best something up to 720p). Camera2 offers much wider range of preview sizes, up to the sensor resolution of the camera. I guess we need an extra parameter or somesuch to select a dedicated frame size...

**Comment by tipsmiller** [comment]
I'm not sure how yours is, but the version of the SDK I downloaded for the last release had an error at [this line](https://github.com/opencv/opencv/blob/d5afa8943b094beef09d3ae6f82d09555b5e6545/modules/java/generator/android-21/java/org/opencv/android/Camera2Renderer.java#L68), where the best size was being forced into the maximum size available. Check to see if your Camera2Renderer is up to date with master.

**Comment by ansanders** [comment]
You are right. Best size was forced to maximum because no other resolution hit the if-condition. 
I saw some examples which calculate the best size with respect to the resulting FPS. But this is only a solution for non legacy devices since you need to get access to the available FPS-Ranges.

**Comment by ansanders** [comment]
@floe I analyzed the behaviour a bit more. 
I increased FPS by ~4 FPS with changing the deliverAndDrawFrame(frame) method of CameraBridgeViewBase. They allocate a lot of memory for rectangle instantiation. So I declared the Rects as class member variables and only update their coordinates in the draw method. 

But now I think it's up to you with another 2 things. 
First of all I dont know why the Android Profiler shows so many instances of the "OpenCvCameraBackground" - Thread. Maybe they are kind of artifacts since I notice FPS drops after closing and restarting the App. Maybe the CameraBackgroundThreads don't get closed  correctly? 

Most CPU Usage of my Nexus 7 goes to your new conversion method : 
`//Imgproc.cvtColorTwoPlane(mYuvFrameData, mUVFrameData, mRgba, Imgproc.COLOR_YUV2RGBA_NV21);
 //Imgproc.cvtColorTwoPlane(mYuvFrameData, mUVFrameData, mRgba, Img proc.COLOR_YUV420sp2BGR);`

Since the upper one gives me a wrong color space (inverted colors) I chose the second one. 
In both cases the CPU usage increases extremly with increasing image resolution. 
I would like to use the optimal resolution but ended up with 960x 720 or so because the framerate dropped too much. 
Are you able to search the wrong behaviour inside this method for me or can we find another approach to fast up the YUV_420 to RGBA (or BGR, since then its not inverted) conversion a bit?

**Comment by JannisW** [comment]
> @tipsmiller Re: [#11261 (comment)](https://github.com/opencv/opencv/issues/11261#issuecomment-380217195)
> 
> I believe [this line](https://github.com/opencv/opencv/blob/d5afa8943b094beef09d3ae6f82d09555b5e6545/modules/java/generator/android-21/java/org/opencv/android/JavaCamera2View.java#L265) from the JavaCamera2View class suffers from the same bug. I think an assumption was made that the StreamConfigurationMap.getOutputSizes method returns a _sorted ascending_ array when in fact it appears that the array returned is sorted descending (though I don't know if that can be relied upon either, as I did not see any documented guarantees about the sort order).

Just as a side note: This bug also affects OpenCV 4.0.1
